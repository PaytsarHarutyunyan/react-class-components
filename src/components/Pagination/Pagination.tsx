import React, { useState, useEffect } from 'react';
import classNames from 'classnames';
import styles from './Pagination.module.css';
import { Link } from 'react-router-dom';
import { useSelectedPage } from '../../hooks/useSelectedPage';

interface PaginationProps {
    pageCount: number;
    action: (page: number) => Promise<void>;
}

const Pagination: React.FC<PaginationProps> = ({ pageCount, action }) => {
    const { getSelectedPage } = useSelectedPage();
    const selectedPage = getSelectedPage();

    const getPageLink = (pageNumber: number) => {
        return `${location.pathname}?page=${pageNumber}`;
    };

    const [paginationItems, setPaginationItems] = useState<number[]>([]);
    const [nextPage, setNextPage] = useState<number>(selectedPage + 1);
    const [prevPage, setPrevPage] = useState<number>(selectedPage - 1);

    useEffect(() => {
        const items = new Array(pageCount).fill(null).map((_, index) => index + 1);
        setPaginationItems(items);
        setNextPage(selectedPage === pageCount ? pageCount : selectedPage + 1);
        setPrevPage(selectedPage === 1 ? selectedPage : selectedPage - 1);
    }, [selectedPage, pageCount]);

    const drawPaginationItems = () =>
        paginationItems.map((paginationItem) => (
            <button
                className={classNames({
                    [styles.item]: true,
                    [styles.itemSelected]: paginationItem === selectedPage,
                })}
                key={paginationItem}
            >
                <Link
                    style={{ display: 'block', color: '#ffffff', textDecoration: 'none' }}
                    onClick={() => action(paginationItem)}
                    to={getPageLink(paginationItem)}
                >
                    {paginationItem}
                </Link>
            </button>
        ));

    return (
        <div className={styles.container}>
            <div className={styles.itemContainer}>
                <button className={styles.item}>
                    <Link
                        style={{ display: 'block', color: '#ffffff', textDecoration: 'none' }}
                        onClick={() => action(1)}
                        to={getPageLink(1)}
                    >
                        &lt;&lt;
                    </Link>
                </button>
                <button className={styles.item}>
                    <Link
                        style={{ display: 'block', color: '#ffffff', textDecoration: 'none' }}
                        onClick={() => action(prevPage)}
                        to={getPageLink(prevPage)}
                    >
                        &lt;
                    </Link>
                </button>
                {paginationItems.length ? drawPaginationItems() : 'No data available'}
                <button className={styles.item}>
                    <Link
                        style={{ display: 'block', color: '#ffffff', textDecoration: 'none' }}
                        onClick={() => action(nextPage)}
                        to={getPageLink(nextPage)}
                    >
                        &gt;
                    </Link>
                </button>
                <button className={styles.item}>
                    <Link
                        style={{ display: 'block', color: '#ffffff', textDecoration: 'none' }}
                        onClick={() => action(pageCount)}
                        to={getPageLink(pageCount)}
                    >
                        &gt;&gt;
                    </Link>
                </button>
            </div>
        </div>
    );
};

export default Pagination;
