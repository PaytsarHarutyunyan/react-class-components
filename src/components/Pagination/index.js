import Footer from './Pagination';
export default Footer;