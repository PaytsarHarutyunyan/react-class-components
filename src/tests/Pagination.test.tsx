import React from 'react';
import { screen, render } from '@testing-library/react';
import Pagination from '../components/Pagination/Pagination';

const mockPush = jest.fn();
jest.mock('next/router', () => ({
    useRouter: jest.fn(() => ({
        pathname: '/',
        query: {},
        push: mockPush,
    })),
}));

jest.mock('next/navigation', () => ({
    useRouter: jest.fn(() => ({
        push: mockPush,
    })),
    redirect: jest.fn(),
    usePathname: jest.fn(() => '/people'),
    useSearchParams: jest.fn(() => ({ get: jest.fn() })),
}));

describe('Pagination Component', () => {
    test('renders pagination and handles page change', () => {
        const { container } = render(<Pagination pageCount={5} />);

        const page2Link = screen.getByText('2');
        expect(page2Link).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });
});
