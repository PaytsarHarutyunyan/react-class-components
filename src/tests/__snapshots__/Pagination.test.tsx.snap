// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination Component renders pagination and handles page change 1`] = `
<div>
  <div
    class="container"
  >
    <div
      class="itemContainer"
    >
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=1"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          &lt;&lt;
        </a>
      </button>
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=NaN"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          &lt;
        </a>
      </button>
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=1"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          1
        </a>
      </button>
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=2"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          2
        </a>
      </button>
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=3"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          3
        </a>
      </button>
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=4"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          4
        </a>
      </button>
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=5"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          5
        </a>
      </button>
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=NaN"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          &gt;
        </a>
      </button>
      <button
        class="item"
        style="position: relative;"
      >
        <a
          href="/people?page=5"
          style="display: flex; color: rgb(255, 255, 255); text-decoration: none; position: absolute; margin: auto; top: 0px; left: 0px; right: 0px; bottom: 0px; align-items: center; justify-content: center;"
        >
          &gt;&gt;
        </a>
      </button>
    </div>
  </div>
</div>
`;
